src/main/java/org/zalando/nakadi/domain/BatchFactory.java

```diff
@@ -1,16 +1,62 @@
 package org.zalando.nakadi.domain;
 
-import org.json.JSONArray;
+import org.json.JSONException;
 
 import java.util.ArrayList;
 import java.util.List;
 
 public class BatchFactory {
 
-    public static List<BatchItem> from(final JSONArray events) {
-        final List<BatchItem> batch = new ArrayList<>(events.length());
-        for (int i = 0; i < events.length(); i++) {
-            batch.add(new BatchItem(events.getJSONObject(i)));
+    public static List<BatchItem> from(final String events) {
+        final List<BatchItem> batch = new ArrayList<>();
+        StringBuilder sb = new StringBuilder();
```

Comment:
I looked into this, and short of using another JSON parser, or doing some ugly size validation inside the controller, I don't see how we can improve this. We should discuss this in more detail.